code/trans.c

```diff
@@ -0,0 +1,377 @@
+ *
+ *
+ * Rationale
+ *
+ * This design was arrived at after some pain.  The MPS isn't really designed
+ * for this kind of thing, and the pools generally assume that they're doing
+ * a garbage collection when they're asked to condemn, scan, fix, and reclaim
+ * stuff.  This makes it very hard to apply the transform without also doing
+ * a garbage collection.  Changing this would require a significant reworking
+ * of the MPS to generalise its ideas, and would bloat the pool classes.
+ *
+ *
+ * Assumptions:
+ *
+ *   - Single-threaded mutator.  In fact this code might work if other mutator
+ *     threads are running, since the shield ought to operate correctly.
+ *     However, in this case there can only be one trace running so that the
+ *     correct fix method is chosen by ScanStateInit.
+ */
+
+#include "trans.h"
+#include "table.h"
+
+
+#define TransformSig         ((Sig)0x51926A45) /* SIGnature TRANSform */
+
+typedef struct mps_transform_s {
+  Sig sig;                      /* <design/sig/> */
+  Arena arena;                  /* owning arena */
+  Table oldToNew;               /* map to apply to refs */
+  Epoch epoch;                  /* epoch in which transform was created */
+  Bool aborted;                 /* no longer transforming, just GCing */
+} TransformStruct;
+
+
+Bool TransformCheck(Transform transform)
+{
+  CHECKS(Transform, transform);
+  CHECKU(Arena, transform->arena);
+  /* .check.boot: avoid bootstrap problem in transformTableAlloc where
+     transformTableFree checks the transform while the table is being
+     destroyed */
+  if (transform->oldToNew != NULL)
+    CHECKD(Table, transform->oldToNew);
+  CHECKL(BoolCheck(transform->aborted));
+  CHECKL(transform->epoch <= ArenaEpoch(transform->arena));
+  return TRUE;
+}
+
+
+/* Allocator functions for the Table oldToNew */
+
+static void *transformTableAlloc(void *closure, size_t size)
+{
+  Transform transform = (Transform)closure;
+  Res res;
+  void *p;
+
+  AVERT(Transform, transform);
+
+  res = ControlAlloc(&p, transform->arena, size);
+  if (res != ResOK)
+    return NULL;
+
+  return p;
+}
+
+static void transformTableFree(void *closure, void *p, size_t size)
+{
+  Transform transform = (Transform)closure;
+  AVERT(Transform, transform);
+  ControlFree(transform->arena, p, size);
+}
+
+
+Res TransformCreate(Transform *transformReturn, Arena arena)
+{
+  Transform transform;
+  Res res;
+  void *p;
+
+  AVER(transformReturn != NULL);
+  AVERT(Arena, arena);
+
+  res = ControlAlloc(&p, arena, sizeof(TransformStruct));
+  if (res != ResOK)
+    goto failAlloc;
+  transform = (Transform)p;
+  
+  transform->oldToNew = NULL;
+  transform->arena = arena;
+  transform->epoch = ArenaEpoch(arena);
+  transform->aborted = FALSE;
+
+  transform->sig = TransformSig;
+
+  AVERT(Transform, transform);
+
+  res = TableCreate(&transform->oldToNew,
+                    0, /* don't grow table until TransformAddOldNew */
```

Comment:
m4. Why not? rule.generic.clear

Reply:
Fixed: Clarified to "no point guessing size before TransformAddOldNew" in 3b97736d8